cmd/tink-worker/cmd/root.go

```diff
@@ -0,0 +1,173 @@
+			if timeOut > 0 {
+				var cancel context.CancelFunc
+				ctx, cancel = context.WithTimeout(ctx, timeOut)
+				defer cancel()
+			}
+
+			conn, err := tryClientConnection(logger, retryInterval, retries)
+			if err != nil {
+				return err
+			}
+			rClient := pb.NewWorkflowSvcClient(conn)
+
+			regConn := internal.NewRegistryConnDetails(registry, user, pwd, logger)
+			worker := internal.NewWorker(rClient, regConn, logger, registry, retries, retryInterval, maxFileSize)
+
+			err = worker.ProcessWorkflowActions(ctx, workerID)
+			if err != nil {
+				return errors.Wrap(err, "worker Finished with error")
+			}
+			return nil
+		},
+	}
+
+	rootCmd.Flags().Duration("retry-interval", retryIntervalDefault, "Retry interval in seconds")
+
+	rootCmd.Flags().Duration("timeout", time.Duration(defaultTimeoutMinutes*time.Minute), "Max duration to wait for worker to complete")
+
+	rootCmd.Flags().Int("max-retry", retryCountDefault, "Maximum number of retries to attempt")
+
+	rootCmd.Flags().Int64("max-file-size", defaultMaxFileSize, "Maximum file size in bytes")
+
+	// rootCmd.Flags().String("log-level", "info", "Sets the worker log level (panic, fatal, error, warn, info, debug, trace)")
+
+	rootCmd.Flags().StringP("id", "i", "", "Sets the worker id")
+	must(rootCmd.MarkFlagRequired("id"))
+
+	rootCmd.Flags().StringP("docker-registry", "r", "", "Sets the Docker registry")
+	must(rootCmd.MarkFlagRequired("docker-registry"))
+
+	rootCmd.Flags().StringP("registry-username", "u", "", "Sets the registry username")
+	must(rootCmd.MarkFlagRequired("registry-username"))
+
+	rootCmd.Flags().StringP("registry-password", "p", "", "Sets the registry-password")
+	must(rootCmd.MarkFlagRequired("registry-password"))
+
+	return rootCmd
+}
+
+// createViper creates a Viper object configured to read in configuration files
+// (from various paths with content type specific filename extensions) and load
+// environment variables that start with TINK_WORKER.
+func createViper() (*viper.Viper, error) {
+	v := viper.New()
+	v.AutomaticEnv()
+	v.SetConfigName("tink-worker")
+	v.AddConfigPath("/etc/tinkerbell")
+	v.AddConfigPath(".")
+	v.SetEnvPrefix("TINK_WORKER")
+	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+
+	// If a config file is found, read it in.
+	if err := v.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
+			return nil, err
+		}
+	} else {
+		fmt.Fprintln(os.Stderr, "Using config file:", v.ConfigFileUsed())
+	}
+
+	return v, nil
+}
+
+func applyViper(v *viper.Viper, cmd *cobra.Command) error {
+	errors := []error{}
+
+	cmd.Flags().VisitAll(func(f *pflag.Flag) {
+		if !f.Changed && v.IsSet(f.Name) {
+			val := v.Get(f.Name)
+			if err := cmd.Flags().Set(f.Name, fmt.Sprintf("%v", val)); err != nil {
+				errors = append(errors, err)
+				return
+			}
+		}
+	})
+
+	if len(errors) > 0 {
+		errs := []string{}
+		for _, err := range errors {
+			errs = append(errs, err.Error())
+		}
+		return fmt.Errorf(strings.Join(errs, ", "))
```

Comment:
I swear I've seen something like multierr in the stdlib and keep thinking we should use it and then never find it again.

Reply:
https://github.com/hashicorp/go-multierror is the one I am aware of

Reply:
https://pkg.go.dev/go.uber.org/multierr is in our go.sum, but not used directly.

Reply:
are you speaking about Wrap/Unwrap @mmlb https://blog.golang.org/go1.13-errors ?

Reply:
@gianarb no, I've used hashicorp multierror and I swear I've thought there one similar in the stdlib. Didn't find one when I went looking, probably just my bad memory.